### PR TITLE
Option to allow all constant scalars to be baked into the graph

### PR DESCRIPTION
**Pull request by cjolivier01**
I know that this seems like a bit of a restrictive option, but could you please allow this option to allow all constant scalars to be "baked" into the graph rather than streaming them in?  I have some use-cases where streaming a lot of values scalars can be problematic for a number of reasons, including network overhead.  I have been using this change in my branch for many months without incident.

Note that this does not affect things such as GetRngSeed

### CODE REVIEW

**Comment by JackCaoG** [comment]
I would put this under the Experiment like 
https://github.com/pytorch/xla/blob/db18ba57137e9e082aab9337e148120ce5725b22/test/cpp/test_aten_xla_tensor.cpp#L4177

since we do not have any test for this. I think this is a "power user option" since it makes recompile happens a lot more easier so we need to be more careful about this. @davidel What do you think?

**Comment by cjolivier01** [comment]
> I would put this under the Experiment like
> 
> https://github.com/pytorch/xla/blob/db18ba57137e9e082aab9337e148120ce5725b22/test/cpp/test_aten_xla_tensor.cpp#L4177
> 
> since we do not have any test for this. I think this is a "power user option" since it makes recompile happens a lot more easier so we need to be more careful about this. @davidel What do you think?
btw, I am happy to write a test for it.  I considered it, but couldn't find a similar test to base it on, i.e. one that set the other environment variable and am not sure what that test would look like.  If you can give some direction tot hat end, i can add it.

**Comment by cjolivier01** [comment]
Or you mean use DebugUtil::ExperimentEnabled("some_word") instead of the environment variable?

**Comment by JackCaoG** [comment]
> Or you mean use DebugUtil::ExperimentEnabled("some_word") instead of the environment variable?

Yes, that's what I meant. One thing you could do is to add a function like

https://github.com/pytorch/xla/blob/26c283987e80564eebffefc433dbc32f9bebf483/test/run_tests.sh#L44

and run our `test_operation` with this "scalar" mode like
https://github.com/pytorch/xla/blob/26c283987e80564eebffefc433dbc32f9bebf483/test/run_tests.sh#L58

**Comment by cjolivier01** [comment]
> Yes, that's what I meant. One thing you could do is to add a function like
> 
> https://github.com/pytorch/xla/blob/26c283987e80564eebffefc433dbc32f9bebf483/test/run_tests.sh#L44
> 
> and run our `test_operation` with this "scalar" mode like
> 
> https://github.com/pytorch/xla/blob/26c283987e80564eebffefc433dbc32f9bebf483/test/run_tests.sh#L58

Ok, I will write this test

**Review by davidel — COMMENTED** [review]
I honestly do not see the reason for this. The scalars are going to be sent once, in parallel, and will never be sent again (unless they change - in which case you are in much worse shape with a recompilation). Scalars are also very small, and with 1VM the already small send time will be close to zero.

**Comment by cjolivier01** [comment]
> I honestly do not see the reason for this. The scalars are going to be sent once, in parallel, and will never be sent again (unless they change - in which case you are in much worse shape with a recompilation). Scalars are also very small, and with 1VM the already small send time will be close to zero.

Hi @davidel !  I have had this option in for some time, before I understood the system as well as I do now, and before the BuildInputOutputAlias function was a added in tensor.cpp and maybe the fundamental behavior by them is misunderstood by me or the behavior has evolved since then.  My understanding is that since these won't show up as aliased input/output parameters, an arbitrary device will not  know that they are intended to persist in memory/on device fabric, As they are at the XLA level, shown simply as graph inputs, so during final device compilation, they will be treated differently (at least in my case), since they seem indistinguishable from a label, which changes for every step.  In addition, it's possible I am missing some fundamental thing here.  As arguments are sent to Execute, is the argument handle the same each time for this scalar, until the value changes (or even after it changes)?  I do note that it changes for the weights (as new handles are returned from Execute, if the result tuple is decomposed, anyway), although maybe that is due to the "update" nature of the weights specifically.

I have another issue but am very interested to know that as it may affect my longer-term strategy.

**Comment by davidel** [comment]
> > I honestly do not see the reason for this. The scalars are going to be sent once, in parallel, and will never be sent again (unless they change - in which case you are in much worse shape with a recompilation). Scalars are also very small, and with 1VM the already small send time will be close to zero.
> 
> Hi @davidel ! I have had this option in for some time, before I understood the system as well as I do now, and before the BuildInputOutputAlias function was a added in tensor.cpp and maybe the fundamental behavior by them is misunderstood by me or the behavior has evolved since then. My understanding is that since these won't show up as aliased input/output parameters, an arbitrary device will not know that they are intended to persist in memory/on device fabric, As they are at the XLA level, shown simply as graph inputs, so during final device compilation, they will be treated differently (at least in my case), since they seem indistinguishable from a label, which changes for every step. In addition, it's possible I am missing some fundamental thing here. As arguments are sent to Execute, is the argument handle the same each time for this scalar, until the value changes (or even after it changes)? I do note that it changes for the weights (as new handles are returned from Execute, if the result tuple is decomposed, anyway), although maybe that is due to the "update" nature of the weights specifically.
> 
> I have another issue but am very interested to know that as it may affect my longer-term strategy.

They will persist in the device data cache, and the handle will be the same (the one returned by the XRT service side when we upload them).

**Comment by cjolivier01** [comment]
> They will persist in the device data cache, and the handle will be the same (the one returned by the XRT service side when we upload them).

Thank you very much for clarifying!  

The problem still arises for me, however, that I do some stuff with the graph, including optimizations and such, and on some devices, where there is not a traditional "memory bus" and samples "flow" down the graph, while weights remain in-place, there is a clear distinction between data which changes (or may change without warning) every step (i.e. features, labels, etc) and data which remains in static memory and isn't easily updated on a step-by-step basis (such as weights, which have this concept defined int he alias map).  At the XLA level, there's no way to tell the difference between such a feature/label and these scalars, which usually don't change, so I would need to treat them as features and physically deliver them every step, which is problematic for what can be an large number of these.  Basically, what II am hoping for is a mode where I can have the equivalent of a tf.Constant where I can just bake it in.

wdyt?

**Comment by davidel** [comment]
> > They will persist in the device data cache, and the handle will be the same (the one returned by the XRT service side when we upload them).
> 
> Thank you very much for clarifying!
> 
> The problem still arises for me, however, that I do some stuff with the graph, including optimizations and such, and on some devices, where there is not a traditional "memory bus" and samples "flow" down the graph, while weights remain in-place, there is a clear distinction between data which changes (or may change without warning) every step (i.e. features, labels, etc) and data which remains in static memory and isn't easily updated on a step-by-step basis (such as weights, which have this concept defined int he alias map). At the XLA level, there's no way to tell the difference between such a feature/label and these scalars, which usually don't change, so I would need to treat them as features and physically deliver them every step, which is problematic for what can be an large number of these. Basically, what II am hoping for is a mode where I can have the equivalent of a tf.Constant where I can just bake it in.
> 
> wdyt?

I see a lot of "in my case" in your argument. The fact that you physically deliver them at every step, means you already have done some major surgeries to the pytorch/xla code, and this can just be one more to it 😉

**Comment by cjolivier01** [comment]
The part I am talking about being delivered is on the other side of the XRT/XLA interface, which won’t know from the XLA it receives (and subsequently compiles) for example the TPU’s XRT server in your case, and not the pytorch/xla code.

**Comment by cjolivier01** [comment]
Maybe I should clarify that there is a second stage of delivery on the other side of the XRT interface, from the stream executor to the actual fabric, and there’s a difference between delivering two values per step and 100,000 values per step. Being impossible to bake a trivial constant into the graph causes unnecessary overhead.

**Comment by davidel** [comment]
So you cannot keep stuff alive on the device? And how can the whole thing work with XRT (or any device interface)?

IMHO this feature does not seem a generally usable one, especially with real life optimizers having step-dependent parametrized computations (LR at the very least), which would constantly cause recompilations.

**Comment by cjolivier01** [comment]
Note that ideally LR and select other things could still be streamed in, as the GetRng seed is, which doesn’t pass through the “special scalar” check.

**Comment by davidel** [comment]
But this is pytorch, not tensorflow, and users are accustomed to do things pytorch has always done.
I am not sure optimizing for users doing things the tensorflow way is a good idea, as all the optimizers I have seen in the pytorch world just use normal variables as per-step computations, not select based ones.
Here I talk about optimizers, but essentially any kind computations which have scalar variables changing across steps, would create damage.

So, again, this is a feature that not only has a very limited use in the real pytorch world, but could create problems if indeed these constants are changing and suddenly users see 30+ seconds halts due to recompilations.

PS: Also this seems to be an issue with your own device (for which - not seeing any OSS code - I assume you have your own tree and you ship your own pytorch version), as all the currently known pytorch/xla devices (CPU, GPU, TPU) do not have the issue due to the device data cache.

**Comment by cjolivier01** [comment]
ok, although the only change in the pytorch version is commenting out a couple of validation lines in one_hot that call item(), which destabilize the graph., which affects all ptxla workflows.
